DR-1386 Improve the performance of creating snapshots

nmalfroy opened the pull request:
After doing a bit of digging, it looks like snapshot creation is currently slowed down by:
- Setting ACLs on files in GCS buckets
- Fetching metadata on the file objects from Firestore

This PR should make these steps significantly faster

For the first point, I added a shared thread pool (which is, by default, bound to a maximum of 50 threads with a max queue size of 1000...there may be some tuning here - this allows running two set acls flight steps at the same time on the same pod before ACL submissions start being rejected) and the ACL commands are distributed in batch across the 50 threads.  On test runs of 2K files, the ACL commands went from taking 5.5 minutes to 20 seconds.

For the second point, I updated the code that fetches file metadata to use the batch metadata fetch method.  In order to do that, I did have to rework how we organize the loop that triggers recursion by splitting the list into directory and file objects, recursing with the directory objects and using batch fetching with the file objects.  Using the same 2K file test, the numbers, the pre-batching time took 4 minutes and went down to 2 seconds with batching.

Note: the PR also includes the TestRunner test that should now be included in the nightly perf test suite

nmalfroy committed:
Add a shared threadpool to use for multithreading tasks

nmalfroy committed:
Add a utility class for working with Futures

...specifically waiting for a list of futures and returning a resolved list of the said futures

nmalfroy committed:
ACL calls against GCS are split into many threads

nmalfroy committed:
Use batched file metadata load

Note: this required splitting the directory enum into directory objects (to recurse over) and file object (to process in bulk)

nmalfroy committed:
In preparation of running longer jobs, make the max job length longer in the tests

nmalfroy committed:
Drive by: log something more useful when deleting a file

nmalfroy committed:
Add a TestRunner test to test perf for creating a snapshot with X files

nmalfroy committed:
Add my own namespace for local testing

nmalfroy committed:
Add CreateSnapshot perf test to nightly test suite

RoriCremer commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -98,7 +107,9 @@ public void deleteFilesFromDataset(Dataset dataset, InterruptibleConsumer<FireSt
         String datasetId = dataset.getId().toString();
+        //Slow
         fileDao.deleteFilesFromDataset(firestore, datasetId, func);
+//        firestore.
```
nit: are comments 110-112 just notes? A little context would be helpful

nmalfroy replied:
Ugh...I missed this when I was reading it over...nice catch.  will remove

RoriCremer commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -165,7 +176,15 @@ public void snapshotCompute(Snapshot snapshot) throws InterruptedException {
             computeDirectory(firestore, snapshotId, topDir, updateBatch);
+            performanceLogger.timerEndAndLog(
+                retrieveTimer,
+                snapshotId, // not a flight, so no job id
```
is there no flight/jobId?

nmalfroy replied:
there didn't seem to be for this method.  I noticed that we do similar things in other calls timerEndAndLog.  I wonder if we really want that second parameter to be flightId or more generally ID

ddietterich commented on datarepo-clienttests/src/main/resources/configs/basicexamples/CreateSnapshot.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "CreateSnapshot",
```
I think this should be in the functional directory not the basicexamples directory.

In fact looking at the whole directory, it looks like there are many files in basic examples that should be in functional.

nmalfroy replied:
Yeah, that's a really good point.  I'll move this there

RoriCremer commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDependencyDao.java:
```diff
@@ -188,7 +188,7 @@ public void deleteSnapshotFileDependencies(Dataset dataset, String snapshotId) t
 
             for (DocumentSnapshot docSnap : batch) {
-                logger.info("deleting: " + docSnap.toString());
+                logger.info("deleting: " + docSnap.getReference().getPath());
```
nice catch

nmalfroy replied:
👍

ddietterich commented on datarepo-clienttests/src/main/java/scripts/testscripts/CreateSnapshot.java:
```diff
@@ -0,0 +1,199 @@
+    // ingest a file -- TODO CannedTestData.getMeA1KBFile
+    URI sourceUri = new URI("gs://jade-testdata/fileloadprofiletest/1KBfile.txt");
+
+    String loadTag = FileUtils.randomizeName("lookupTest");
```
Is there are reason you could not adapt `datarepo-clienttests/src/main/java/scripts/utils/BulkLoadUtils.java` to do the loads?

nmalfroy replied:
Only because I didn't know it existed :). Will update

ddietterich commented on datarepo-clienttests/src/main/java/scripts/testscripts/CreateSnapshot.java:
```diff
@@ -0,0 +1,199 @@
+      }
+    }
+    String jsonLines;
+    try (Stream<String> resultStream = fileIds.stream()) {
```
This is a lot of complexity for a simple thing. A try block, stream creation, mapping and collector to look through an array and generate a String array only to turn it into a byte string. Here is my way:

```
StringBuilder sb = new StringBuilder();
for (String fileId : fileIds) {
  sb.append("{\"VCF_File_Name\":\"name1\", \"Description\":\"description1\", \"VCF_File_Ref\":\""
                          + fileId
                          + "\"}\n");
}
byte[] fileRefBytes = sb.toString().getBytes();
```

nmalfroy replied:
Yeah, also, I don't think I even need to do it if I use the method from BulkLoad Utils.  Will remove (actually, just reworking the setup method to look like `BuildSnapshotWithFiles.java`

fboulnois replied:
Also, perhaps this method could be split up into smaller units? It's not clear how much state is actually being carried across this function.

nmalfroy replied:
Yeah, it's basically loading a dataset with files.  This should get muuuch simpler once we have a way to reference files in the original data being loaded rather than needing scratch files.  One reason I did end up having to do a bunch of this is to chunk since the load files endpoints takes at most 1000 files.  I'll move that logic into BulkLoadUtils

pshapiro4broad commented on datarepo-clienttests/src/main/java/scripts/testscripts/CreateSnapshot.java:
```diff
@@ -0,0 +1,199 @@
+
+  /** Public constructor so that this class can be instantiated via reflection. */
+  public CreateSnapshot() {
+    super();
```
`super();` unnecessary as Java inserts the call for you

pshapiro4broad commented on datarepo-clienttests/src/main/java/scripts/testscripts/CreateSnapshot.java:
```diff
@@ -0,0 +1,199 @@
+public class CreateSnapshot extends SimpleDataset {
+  private static final Logger logger = LoggerFactory.getLogger(CreateSnapshot.class);
+
+  /** Public constructor so that this class can be instantiated via reflection. */
```
I wouldn't use javadoc for this type of comment since it's an implementation detail and not information for a client of this class.

pshapiro4broad commented on datarepo-clienttests/src/main/java/scripts/testscripts/CreateSnapshot.java:
```diff
@@ -0,0 +1,199 @@
+
+  private SnapshotModel snapshotModel;
+
+  private static List<BlobId> scratchFiles = new ArrayList<>();
```
Is `static` required here? Using it means that the code needs to be `synchronized` at some point to be thread-safe.

Also, this can be `final`.

ddietterich commented on src/main/java/bio/terra/common/FutureUtils.java:
```diff
@@ -0,0 +1,76 @@
+            List<T> returnList = stream
+                .map(f -> {
+                    try {
+                        if (maxThreadWait.isPresent()) {
```
Is there any benefit to performing the get - and perhaps waiting - if a failure has already been found?
It is optimizing the failure path, so maybe no worth it. OTOH, if you are collecting stuck operation errors by waiting maxThreadWait for each, it could get to be a long time.

nmalfroy replied:
Great question and I was a little torn about it.  I actually started with it doing that but I liked the idea of waiting for all threads to run their course.  I guess one thought would be to cancel all remaining jobs as soon as a failure is found.  As I type it out loud, that may actually make more sense.

nmalfroy replied:
I'm going to update to make it cancel all of the outstanding tasks

fboulnois replied:
👍  on cancelling the other tasks.

nmalfroy replied:
Yeah.  I guess I had been thinking of keeping similar to how JS handles bulk waiting for futures but this definitely seems like a better approach

ddietterich commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -98,7 +107,9 @@ public void deleteFilesFromDataset(Dataset dataset, InterruptibleConsumer<FireSt
         DatasetDataProject dataProject = dataLocationService.getProjectOrThrow(dataset);
         Firestore firestore = FireStoreProject.get(dataProject.getGoogleProjectId()).getFirestore();
         String datasetId = dataset.getId().toString();
+        //Slow
```
Are these comments still useful? They don't mean anything to me.

nmalfroy replied:
Sorry, nope. That was a beacon to myself sine I was getting lost in the code. Missed it in my sweep

pshapiro4broad commented on datarepo-clienttests/src/main/java/scripts/testscripts/CreateSnapshot.java:
```diff
@@ -0,0 +1,199 @@
+            DataRepoUtils.expectJobSuccess(
+                repositoryApi, ingestFileJobResponse, BulkLoadArrayResultModel.class);
+
+        try (Stream<BulkLoadFileResultModel> resultStream =
```
I know `Stream` is `AutoCloseable` but I don't think I've seen non-I/O code use try-with-resources to ensure that they're closed. Is there a reason to do that here?

nmalfroy replied:
Yeah, I've been pretty gung-ho about doing this wherever we use streams since my experience has been, if this isn't standard practice, you tend to get unclosed resources that pop up (since you need to constantly rationalize when a stream should or shouldn't be closed....best practice I think is to just always make sure it's closed.

All that being said, I would love for us to use a library like eclipse collections which let's you get rid of 99% of uses of `Stream` instances

pshapiro4broad commented on datarepo-clienttests/src/main/java/scripts/testscripts/CreateSnapshot.java:
```diff
@@ -0,0 +1,199 @@
+        try (Stream<BulkLoadFileResultModel> resultStream =
+            bulkLoadArrayResultModel.getLoadFileResults().stream()) {
+          fileIds.addAll(
+              resultStream.map(BulkLoadFileResultModel::getFileId).collect(Collectors.toList()));
```
You can avoid creating a collection here by doing an `add` in the streaming operation
```suggestion
            resultStream.map(BulkLoadFileResultModel::getFileId).forEach(fileIds::add);
```

ddietterich commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -408,24 +427,56 @@ FireStoreDirectoryEntry computeDirectory(
+                    CollectionUtils.collect(entry.getValue(), dirItem -> {
+                        final FireStoreFile file = fireStoreFilesById.get(dirItem.getFileId());
+                        if (file == null) {
+                            throw new ApiException("File metadata was missing");
```
Ditto, re exception

nmalfroy replied:
👍

ddietterich commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -408,24 +427,56 @@ FireStoreDirectoryEntry computeDirectory(
+                    try {
+                        return computeDirectory(firestore, snapshotId, f, updateBatch);
+                    } catch (InterruptedException e) {
+                        throw new ApiException("Error computing directory metadata", e);
```
We generally use specific exceptions that derive from base class exceptions in common/exceptions. Those exceptions are associated with Http status results as well. For example, `bio/terra/service/filedata/exception/FileNotFoundException.java`

nmalfroy replied:
ok, good to know

nmalfroy replied:
I guess in this case for an interrupt, I wasn't sure what common exception we would have for that so ApiException looked reasonable

ddietterich replied:
I think this should come back to the client as a 500, so i would make an exception that derived from InternalServerErrorException.

nmalfroy replied:
👍

ddietterich commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -408,24 +427,56 @@ FireStoreDirectoryEntry computeDirectory(
+                final List<FireStoreFile> fireStoreFiles =
+                    fileDao.batchRetrieveFileMetadata(firestore, entry.getKey(), entry.getValue());
+
+                // Index by fileId to be able to look up
```
Why can't you just put the `dirItem` into the `enumComputed` here instead of building a separate structure just to re-look it up?

nmalfroy replied:
That's kind of the crux of the change here.  We were able do that before because we were looking up the metadata 1 by 1 (which was the slow bit).  This goal here was to batch but the problem is that it adds complexity because you need to resolve the results afterwards.  So I get the bulk results from the metadata look up but then I need to take the original `FireStoreDirectoryEntry` objects and look up what was returned from firebase to augment them with the appropriate size and metadata.

ddietterich replied:
I understand the batching idea, just not the need to build the correlation map below this line. The batchRetrieve output list is in exactly the order of the input list. So I think you can just walk the entry.getValue() list in parallel with the fireStoreFiles list rather than constructing a map to correlate them.

nmalfroy replied:
Ah interesting.  That's a really good point.  I generally get nervous about positional dependency (since nothing stops somebody from changing the underlying implementation to something non-order-maintaining) but that makes sense with the current way the code is and saves us having to build up this separate map.   I'll maybe add a comment to `batchRetrieveFileMetadata` to make it clear that order in = order out

ddietterich replied:
The documentation on that could certainly use a fix. We depend on that order in = order out property in most of the places where we use it. And it is coded to maintain that property.

nmalfroy replied:
Totally.  That's really good to know.  I'll update

ddietterich commented on src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:
```diff
@@ -48,19 +55,25 @@
         ApplicationContext applicationContext,
-        ConfigurationService configurationService) {
+        ConfigurationService configurationService,
+        @Qualifier("performanceThreadpool") ExecutorService executor,
```
Huh, I've never had to use @Qualifier before. This is needed because you wanted to call it `executor`?

nmalfroy replied:
Yeah, I put it in pre-emptively (perhaps that's confusing) since we may potentially have multiple Executor services.  This let's us get this one specifically.  I can remove for now since we probably don't need it yet

ddietterich replied:
I wouldn't change it, since we could get weird errors if we add more Executor services in the future.
Since I hadn't seen it before, I wanted to make sure I understood what it was.

nmalfroy replied:
👍

ddietterich commented on src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:
```diff
@@ -256,43 +269,56 @@ private void fileAclOp(AclOp op, Dataset dataset, List<String> fileIds, Map<IamR
+
+            try (Stream<FSFile> stream = files.stream()) {
+                List<Future<FSFile>> futures = stream.map(file -> executor.submit(() -> {
+                    // Cache the bucket resources to avoid repeated database lookups
```
I think this would be more readable if you pulled out the unnamed lambda into a method.

nmalfroy replied:
Fair enough.

ddietterich commented on src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:
```diff
@@ -256,43 +269,56 @@ private void fileAclOp(AclOp op, Dataset dataset, List<String> fileIds, Map<IamR
         }
 
-        Map<String, GoogleBucketResource> bucketCache = new HashMap<>();
+        Map<String, GoogleBucketResource> bucketCache = new ConcurrentHashMap<>();
```
Maybe comment what String will be holding in this map

nmalfroy replied:
👍

ddietterich commented on src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:
```diff
@@ -256,43 +269,56 @@ private void fileAclOp(AclOp op, Dataset dataset, List<String> fileIds, Map<IamR
+            try (Stream<FSFile> stream = files.stream()) {
+                List<Future<FSFile>> futures = stream.map(file -> executor.submit(() -> {
+                    // Cache the bucket resources to avoid repeated database lookups
+                    GoogleBucketResource bucketForFile = bucketCache.get(file.getBucketResourceId());
```
Maybe wrap the bucket cache into its own nested class that would hold the concurrent map and perform the test?

nmalfroy replied:
Oh wait, there's maybe a way to make this a bit more concise with a `computeIfAbsent`.  It's standard `Map` functionality so I don't know if it warrants a new class

ddietterich replied:
I don't know if a class is warranted either - just crossed my mind as I was reading through it.

nmalfroy replied:
Also just noticed a thing with that and will a `synchronize` around the `computeIfAbsent` method since there's a race condition there

ddietterich reviewed (changes requested):
A few changes I'd like you to consider.

nmalfroy committed:
Make it so that all outstanding tasks are canceled when one fails

nmalfroy committed:
Simplify TestRunner test and move it to functional

nmalfroy committed:
Add some comments about variaous utility methods

nmalfroy committed:
Split ACL lambda into it's own method and make it block on accessing cache

nmalfroy committed:
Followons for FireStoreDao

- Add more meaningful exception class
- Throw more meaningful exception classes
- Don't create new structure just to join looked up file with files being returned
-

fboulnois reviewed (commented):
I agree with a few of the earlier comments. A few additional suggestions:

nmalfroy committed:
Increase batch size for uploading files in the CreateSnapshot test

nmalfroy committed:
Make FutureUtilsTest more resilient

pshapiro4broad commented on datarepo-clienttests/src/main/java/scripts/testscripts/CreateSnapshot.java:
```diff
@@ -0,0 +1,156 @@
+    logger.info("Successfully loaded data into dataset: {}", ingestResponse.getDataset());
+  }
+
+  private int filesToLoad;
```
Not sure what the Jade conventions are, but I'm used to seeing fields declared together at the top of a class.

nmalfroy replied:
Yeah, that's a great point. I'll move this

pshapiro4broad commented on datarepo-clienttests/src/main/java/scripts/testscripts/CreateSnapshot.java:
```diff
@@ -0,0 +1,156 @@
+    if (parameters == null || parameters.size() == 0) {
+      throw new IllegalArgumentException(
+          "Must provide a number of files to load in the parameters list");
+    } else {
```
unnecessary `else` after throw

nmalfroy replied:
Agreed.  I've just alway found explicit else's a little more readable

pshapiro4broad commented on datarepo-clienttests/src/main/java/scripts/utils/DataRepoUtils.java:
```diff
@@ -37,7 +37,8 @@
   private DataRepoUtils() {}
 
-  private static int maximumSecondsToWaitForJob = 500;
+  private static int maximumSecondsToWaitForJob =
```
Can be `final`. Also, do we not use all uppercase for static final constants?

nmalfroy replied:
My experience is that static + final = all caps.  That being said, yes this should be final

pshapiro4broad commented on datarepo-clienttests/src/main/java/scripts/utils/DataRepoUtils.java:
```diff
@@ -37,7 +37,8 @@
 
-  private static int maximumSecondsToWaitForJob = 500;
+  private static int maximumSecondsToWaitForJob =
+      Long.valueOf(TimeUnit.HOURS.toSeconds(2)).intValue();
```
Is this preferable to writing
```suggestion
      (int) TimeUnit.HOURS.toSeconds(2);
```

ddietterich commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -408,24 +426,47 @@ FireStoreDirectoryEntry computeDirectory(
+                final List<FireStoreFile> fireStoreFiles =
+                    fileDao.batchRetrieveFileMetadata(firestore, entry.getKey(), entry.getValue());
+
+                final AtomicInteger index = new AtomicInteger(0);
```
Why is atomic needed here?

nmalfroy replied:
Ah, that's because since I'm using index in a lambda, it needs to be effectively final.  I tend to like using `AtomicIntegers` for this.  I could have some something like storing it in a one element integer array I just fund this easier to work with

ddietterich reviewed (approved):
lgtm - one question for my own education

pshapiro4broad commented on datarepo-clienttests/src/main/java/scripts/testscripts/CreateSnapshot.java:
```diff
@@ -0,0 +1,156 @@
+      // now go and retrieve the file Id that should be stored in the snapshot
+      snapshotModel = repositoryApi.retrieveSnapshot(snapshotSummaryModel.getId());
+    } catch (Exception e) {
+      logger.error("Error in journey", e);
```
Is this needed for a test? I would assume that the top level catch will also dump the exception and a full stack trace, so you'll know exactly where the exception was raised. In general I prefer logging at the catch site, not the throw; as there are typically many throws but few top level catches.

pshapiro4broad commented on src/main/java/bio/terra/app/configuration/ApplicationConfiguration.java:
```diff
@@ -122,6 +127,21 @@
+     * Certain operations can be spread to run asynchronously to gain a performance boost.  Instead of having
+     * each such task create its own threadpool, this property is used to create a globally accessible pool that should
+     * be used by all such operations.
+     *
```
If you want a paragraph break in a javadoc, you need a `<p>`
```suggestion
     * <p>
```

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -408,24 +426,47 @@ FireStoreDirectoryEntry computeDirectory(
+
+        // Recurse to compute results from underlying directories
+        try (Stream<FireStoreDirectoryEntry> stream = enumDir.stream()) {
+            enumComputed.addAll(stream
```
In general, I think `collection.addAll(stream...collect(Collectors.toList())` is more clearly written as `stream...foreach(collection::add)`. This also avoid creating another collection.

Although in this case `enumComputed` is empty so you can initialize it here instead of using `addAll`

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -408,24 +426,47 @@ FireStoreDirectoryEntry computeDirectory(
-                enumComputed.add(computeDirectory(firestore, snapshotId, dirItem, updateBatch));
+
+        // Recurse to compute results from underlying directories
+        try (Stream<FireStoreDirectoryEntry> stream = enumDir.stream()) {
```
Why are you using try-with-resources to ensure that this stream is closed? I don't think I've seen this idiom before.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -408,24 +426,47 @@ FireStoreDirectoryEntry computeDirectory(
+                final AtomicInteger index = new AtomicInteger(0);
+                enumComputed.addAll(
+                    CollectionUtils.collect(entry.getValue(), dirItem -> {
+                        final FireStoreFile file = fireStoreFiles.get(index.getAndIncrement());
```
So this depends on `batchRetrieveFileMetadata` returning results in the same order that the arguments are in. Maybe worth adding a comment for?

pshapiro4broad commented on src/main/java/bio/terra/common/FutureUtils.java:
```diff
@@ -0,0 +1,79 @@
+     * order of the passed in futures
+     */
+    public static <T> List<T> waitFor(final List<Future<T>> futures,
+                                      final Optional<Duration> maxThreadWait) {
```
What is the benefit of using `Optional<>` over the object itself in this case? As I understand it, the intended use of `Optional` is as a function return value. For an argument, using `Duration` and allowing `null` provides the same feature.

pshapiro4broad commented on src/main/java/bio/terra/common/FutureUtils.java:
```diff
@@ -0,0 +1,79 @@
+    public static <T> List<T> waitFor(final List<Future<T>> futures,
+                                      final Optional<Duration> maxThreadWait) {
+
+        final AtomicReference<Optional<ApiException>> foundFailure = new AtomicReference<>(Optional.empty());
```
Also here, using `ApiException` directly seems more straightforward to me.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:
```diff
@@ -256,43 +270,74 @@ private void fileAclOp(AclOp op, Dataset dataset, List<String> fileIds, Map<IamR
+            final URI gsUri = URI.create(file.getGspath());
+            final String bucketPath = StringUtils.removeStart(gsUri.getPath(), "/");
+            final BlobId blobId = BlobId.of(bucketForFile.getName(), bucketPath);
+            switch (op) {
```
Whenever I see a `switch` over an `enum` like this, I think about rewriting it using enum methods. This avoids the need for `default` and makes the calling code cleaner. I wrote it up like this:

```
    private enum AclOp {
        ACL_OP_CREATE {
            void handle(Storage storage, BlobId blobId, List<Acl> acls, List<Acl.Group> groups) {
                for (Acl acl : acls) {
                    storage.createAcl(blobId, acl);
                }
            }
        },
        ACL_OP_DELETE {
            void handle(Storage storage, BlobId blobId, List<Acl> acls, List<Acl.Group> groups) {
                for (Acl.Group group : groups) {
                    storage.deleteAcl(blobId, group);
                }
            }
        };

        abstract void handle(Storage storage, BlobId blobId, List<Acl> acls, List<Acl.Group> groups);
    }
```
but maybe it would read better if the rest was moved into `AclOp` as well.

pshapiro4broad reviewed (commented):
OK, here's my collection of nits and comments. Let me know if there's anything you'd like to talk about in more detail.
